lib/compat/wordpress-5.9/class-gutenberg-rest-templates-controller.php

```diff
@@ -332,6 +332,61 @@ public function delete_item( $request ) {
 		return $this->prepare_item_for_response( $template, $request );
 	}
 
+	/**
+	 * Requesting this endpoint for a template like "twentytwentytwo//home" requires using
+	 * a path like /wp/v2/templates/twentytwentytwo//home. There are special cases when
+	 * WordPress routing corrects the name to contain only a single slash like "twentytwentytwo/home".
+	 *
+	 * This method attempts to find a template with a specific ID, and if it's missing then it
+	 * falls back to parsing REQUEST_URI in an attempt to grab the verbatim ID passed by the user.
+	 *
+	 * See https://core.trac.wordpress.org/ticket/54507 for more context
+	 *
+	 * @param string  $id        ID of the template.
+	 * @param boolean $from_file Optional Whether to use gutenberg_get_block_template instead of get_block_file_template.
+	 * @return WP_Block_Template|null Template.
+	 */
+	private function get_template_by_id( $id, $from_file = false ) {
+		$getter   = $from_file ? 'get_block_file_template' : 'gutenberg_get_block_template';
+		$template = $getter( $id, $this->post_type );
+		if ( ! $template ) {
+			// Make sure REQUEST_URI is set.
+			if ( empty( $_SERVER['REQUEST_URI'] ) ) {
+				return;
+			}
+
+			$uri       = $_SERVER['REQUEST_URI'];
+			$delimiter = $this->namespace . '/' . $this->rest_base . '/';
+
+			// Extract the part of the path that comes after /wp/v2/templates/.
+			$start = strpos( $uri, $delimiter );
+			if ( false === $start ) {
+				return;
+			}
```

Comment:
Even though it's unlikely that `$delimiter` should match anything other than what we seek, could we be stricter about this and reconstruct the entire URI up to the inferred ID? Loosely:

```php
if ( 0 !== strpos( $uri, $prefix ) ) return;
```

Reply:
What would be `$prefix` in this scenario? I think we could just check for multiple matches, e.g. 

```php
if ( false !== strpos( $uri, $delimiter, $start + strlen( $delimiter ) ) ) {
    return;
}
```